sdk/log/logger_test.go

```diff
@@ -273,3 +273,38 @@ func TestLoggerEnabled(t *testing.T) {
 		})
 	}
 }
+
+func TestAllocationLimits(t *testing.T) {
+	const runs = 1
+
+	logger := newLogger(NewLoggerProvider(), instrumentation.Scope{})
+
+	r := log.Record{}
+	r.SetTimestamp(time.Date(2000, time.January, 1, 0, 0, 0, 0, time.UTC))
+	r.SetObservedTimestamp(time.Date(2000, time.January, 1, 0, 0, 0, 0, time.UTC))
+	r.SetBody(log.StringValue("testing body value"))
+	r.SetSeverity(log.SeverityInfo)
+	r.SetSeverityText("testing text")
+
+	attrs5 := []log.KeyValue{
+		log.String("k1", "str"),
+		log.Float64("k2", 1.0),
+		log.Int("k3", 2),
+		log.Bool("k4", true),
+		log.Bytes("k5", []byte{1}),
+	}
+	r.AddAttributes(attrs5...)
+
+	r10 := r
+	r10.AddAttributes(attrs5...)
+	assert.Equal(t, 10, r10.AttributesLen())
+
+	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+		logger.newRecord(context.Background(), r)
+	}), "newRecord")
+
+	// TODO: Optimize this allocation count to 1.
+	assert.Equal(t, 8.0, testing.AllocsPerRun(runs, func() {
```

Comment:
I might miss something; I don't know why I got `4` without the `-race` flag. Any ideas?

Reply:
It executes

```go
	r.back = slices.Grow(r.back, len(attrs[i:]))
	r.back = append(r.back, attrs[i:]...)
```

5 times for attributes from 6th to 10th.

Because of

```go
	r.WalkAttributes(func(kv log.KeyValue) bool {
		newRecord.AddAttributes(kv)
		return true
	})
```

We can consider doing something more or less like this for `r.AttributesLen() > X` (where X is probably something like 6).

```
	n := r.AttributesLen()
	kvs := make([]log.KeyValue, 0, n)
	r.WalkAttributes(func(kv log.KeyValue) bool {
		kvs = append(kvs, kv)
		newRecord.AddAttributes(kv)
		return true
	})
	newRecord.AddAttributes(kvs...)
```

We would need to have benchmarks to find a good value for `X`. 

PS. I am not sure if this is a performance improvement worth the investment given that according to https://youtu.be/tC4Jt3i62ns?si=qO_GGBmtayeLJ5fy&t=367 ~95% of code uses up to 5 attributes.

Reply:
I mean, when I ran the test with the `-race` flag, I got `8`. Without the `-race` flag, I got `4`.

Reply:
`-race` adds some stuff for sake of the race detector and removes some optimizations as well

Reply:
We probably do not want this since it is ultimately testing the allocation strategy for slices of the compiler used.